easybuild/tools/containers.py

```diff
@@ -0,0 +1,305 @@
+    if bootstrap_agent not in SINGULARITY_BOOTSTRAP_TYPES:
+        known_bootstrap_agents = ', '.join(SINGULARITY_BOOTSTRAP_TYPES)
+        raise EasyBuildError("Bootstrap agent in container base spec must be one of: %s" % known_bootstrap_agents)
+
+    res = {'bootstrap_agent': bootstrap_agent}
+
+    for idx, base_spec in enumerate(base_specs[1:]):
+        res.update({'arg%d' % (idx + 1): base_spec})
+
+    return res
+
+
+def generate_singularity_recipe(easyconfigs, container_base):
+    """Main function to Singularity definition file and image."""
+
+    cont_path = container_path()
+
+    # make sure location to write container recipes & images exists
+    mkdir(cont_path, parents=True)
+
+    base_specs = parse_container_base(container_base)
+
+    # extracting application name,version, version suffix, toolchain name, toolchain version from
+    # easyconfig class
+
+    bootstrap_agent = base_specs['bootstrap_agent']
+
+    base_image, base_image_tag = None, None
+
+    # with localimage it only takes 2 arguments. --container-base localimage:/path/to/image
+    # checking if path to image is valid and verify image extension is '.img' or '.simg'
+    if base_specs['bootstrap_agent'] == LOCALIMAGE:
+        base_image = base_specs['arg1']
+        if os.path.exists(base_image):
+            # get the extension of container image
+            image_ext = os.path.splitext(base_image)[1]
+            if image_ext == '.img' or image_ext == '.simg':
+                _log.debug("Extension for base container image to use is OK: %s", image_ext)
+            else:
+                raise EasyBuildError("Invalid image extension '%s' must be .img or .simg", image_ext)
+        else:
+            raise EasyBuildError("Singularity base image at specified path does not exist: %s", base_image)
+
+    # otherwise, bootstrap agent is 'docker' or 'shub'
+    # format --container-base {docker|shub}:<image>:<tag>
+    else:
+        base_image = base_specs['arg1']
+        # image tag is optional
+        base_image_tag = base_specs.get('arg2', None)
+
+    bootstrap_from = base_image
+    if base_image_tag:
+        bootstrap_from += ':' + base_image_tag
+
+    # if there is osdependencies in easyconfig then add them to Singularity recipe
+    install_os_deps = ''
+    for ec in easyconfigs:
+        for osdep in ec['ec']['osdependencies']:
+            if isinstance(osdep, basestring):
+                install_os_deps += "yum install -y %s\n" % osdep
+            # tuple entry indicates multiple options
+            elif isinstance(osdep, tuple):
+                install_os_deps += "yum --skip-broken install %s\n" % ' '.join(osdep)
+            else:
+                raise EasyBuildError("Unknown format of OS dependency specification encountered: %s", osdep)
+
+    # module names to load in container environment
+    mod_names = [e['ec'].name + '/' + det_full_ec_version(e['ec']) for e in easyconfigs]
```

Comment:
Is this necessary? Correct me if I'm wrong but I think the module creation step of the build takes care of loading the depended on modules.

Reply:
@FooBarQuaxx This is not related to dependencies, we only construct a list of module names for the easyconfigs that were specified on the command line here.

We can simplify this to use `e['ec'].full_mod_name` though, so I'll make that change.

Reply:
i agree with @FooBarQuaxx eb just needs to load the module for the calling application and eb will load all the dependent modules. This is true for Flat Naming Scheme since all modules are present in one moduletree. 
I think we can remove the logic and references for variable ``mod_names``

Reply:
@shahzebsiddiqui No, how will we know which modules to load in the `%environment` section of the Singularity image? We need to module names for this...